File: src/packages/control-ui/src/api/subscription_pipeline.ts
export interface IngestedProxyNode {
  protocol: string;
  endpoint: string;
  tag: string;
}

export function parseSubscriptionPayload(content: string): IngestedProxyNode[] {
  let decoded = content.trim();
  try {
    // Attempt base64 decode
    if (typeof atob === 'function') {
      decoded = atob(decoded);
    } else {
      const buf = nodeGlobalBuffer();
      if (buf) {
        decoded = buf.from(decoded, 'base64').toString('utf-8');
      }
    }
  } catch {
    // Fallback to plain string
  }

  const lines = decoded.split('\n');
  const nodes: IngestedProxyNode[] = [];
  const seen = new Set<string>();

  for (const rawLine of lines) {
    const line = rawLine.trim();
    if (!line || line.startsWith('#') || !line.includes('://')) continue;

    const [protocol, restWithTag] = line.split('://', 2);
    let endpoint = restWithTag;
    let tag = '';
    if (restWithTag?.includes('#')) {
      const parts = (restWithTag ?? '').split('#', 2);
      endpoint = parts[0];
      tag = parts[1]!;
    }

    if (!seen.has(line)) {
      seen.add(line);
      nodes.push({ protocol: protocol!, endpoint: endpoint!, tag });
    }
  }

  return nodes;
}

File: labs/mobile/donor-archive/tunnel/SubscriptionPipeline.kt
package com.luminet.android.tunnel

data class ParsedSubscriptionNode(
    val protocol: String,
    val endpoint: String,
    val tag: String
)

class SubscriptionPipeline {
    fun parse(content: String): List<ParsedSubscriptionNode> {
        val lines = content.trim().split("\n")
        val nodes = mutableListOf<ParsedSubscriptionNode>()
        val seen = mutableSetOf<String>()

        for (line in lines) {
            val trimmed = line.trim()
            if (trimmed.isBlank() || trimmed.startsWith("#")) continue
            if (!trimmed.contains("://")) continue

            val parts = trimmed.split("://", limit = 2)
            val protocol = parts[0]
            var rest = parts[1]
            var tag = ""

            if (rest.contains("#")) {
                val tagParts = rest.split("#", limit = 2)
                rest = tagParts[0]
                tag = tagParts[1]
            }

            if (seen.add(trimmed)) {
                nodes.add(ParsedSubscriptionNode(protocol, rest, tag))
            }
        }
        return nodes
    }
}
